src/main/scala/com/microsoft/hyperspace/actions/OptimizeAction.scala

```diff
@@ -0,0 +1,127 @@
+import org.apache.spark.sql.SparkSession
+
+import com.microsoft.hyperspace.HyperspaceException
+import com.microsoft.hyperspace.index._
+import com.microsoft.hyperspace.index.DataFrameWriterExtensions.Bucketizer
+import com.microsoft.hyperspace.index.IndexConstants.OPTIMIZE_MODES
+import com.microsoft.hyperspace.telemetry.{AppInfo, HyperspaceEvent, OptimizeActionEvent}
+import com.microsoft.hyperspace.util.{HyperspaceConf, PathUtils}
+
+/**
+ * Optimize Action provides an optimize support for indexes where small index files
+ * can be merged into larger ones for better index performance.
+ *
+ * Algorithm outline:
+ * 1. Collect all the currently valid index files.
+ * 2. Split files into small and large, based on a threshold.
+ * 3. Bucketwise combine smaller files into 1 file per bucket.
+ * 4. Update index snapshot to remove small files and keep large files + newly created files.
+ *
+ * NOTE: This is an index-only operation. It does not look at the current state of
+ * the data at all. If the data was changed after index creation, optimize will NOT
+ * include the data changes.
+ *
+ * Available modes:
+ * `Quick` mode: This mode allows for fast optimization. Files smaller than a
+ * predefined threshold "spark.hyperspace.index.optimize.fileSizeThreshold" will be picked
+ * for compaction.
+ *
+ * `Full` mode: This allows for slow but complete optimization. ALL files are
+ * eligible for compaction. Compaction threshold is inifinity.
+ *
+ * TODO: Optimize can be a no-op if there is only one small file per bucket.
+ *  https://github.com/microsoft/hyperspace/issues/204.
+ *
+ * @param spark SparkSession
+ * @param logManager IndexLogManager for index being refreshed
+ * @param dataManager Index DataManager for index being refreshed
+ * @param mode Acceptable optimize modes are `quick` and `full`.
+ */
+class OptimizeAction(
+    spark: SparkSession,
+    logManager: IndexLogManager,
+    dataManager: IndexDataManager,
+    mode: String)
+    extends RefreshActionBase(spark, logManager, dataManager) {
+  final override def op(): Unit = {
+    // Rewrite index from small files.
+    val numBuckets = previousIndexLogEntry.numBuckets
+    val indexDF = spark.read.parquet(smallFiles.map(_.name): _*)
+
+    val repartitionedDf =
+      indexDF.repartition(numBuckets, indexConfig.indexedColumns.map(indexDF(_)): _*)
+
+    repartitionedDf.write.saveWithBuckets(
+      repartitionedDf,
+      indexDataPath.toString,
+      logEntry.asInstanceOf[IndexLogEntry].numBuckets,
+      indexConfig.indexedColumns)
+  }
+
+  override def validate(): Unit = {
+    super.validate()
+
+    if (!OPTIMIZE_MODES.exists(_.equalsIgnoreCase(mode))) {
+      throw HyperspaceException(s"Unsupported optimize mode '$mode' found.")
+    }
+
+    if (smallFiles.isEmpty) {
+      throw NoChangesException(
+        s"Optimize aborted as no index files smaller than " +
+          s"${HyperspaceConf.optimizeFileSizeThreshold(spark)} found.")
+    }
+  }
```

Comment:
For the full mode, what if the index is already optimized?

Reply:
Thanks @imback82 , there is a TODO and issue link in the class documentation.
https://github.com/microsoft/hyperspace/pull/166/files/ebfb99a615dfda18b8b9f435054109b72cf74d2f#diff-426163ac5a5a9239dfa09d8f47f0f052f3f194cefd648fe980108d7690827d42R53

I will update the PR description

Reply:
thanks!